url46a88cc932_toby1984_tinyc_tgz-pJ8-de_codesourcery_dcpu16_compiler_CompilerJ8/src/de/codesourcery/dcpu16/parser/Lexer.java

```diff
@@ -0,0 +1,433 @@
+            
+            final InputStream newIn = createInputStream( currentOffset );
+            Lexer.this.scanner = new Scanner( newIn, currentOffset);
+        }
+    }
+    
+    private InputStream createInputStream(int offset) throws EOFException, IOException {
+        return advanceInputStream( input.createInputStream() , offset );
+    }
+    
+    private static InputStream advanceInputStream(InputStream newIn, int bytesToSkip) throws EOFException, IOException 
+    {
+        int toAdvance = bytesToSkip;
+        while ( toAdvance > 0 ) 
+        {
+            if ( newIn.read() == -1 ) {
+                throw new EOFException();
+            }
+            toAdvance--;
+        }
+        return newIn;
+    }
+    
+    public String toString() {
+    	return tokens.toString();
+    }
+    
+    public Lexer(IInputStreamProvider input,boolean ignoreWhitespace) throws IOException 
+    {
+        this.input = input;
+        this.scanner = new Scanner(input.createInputStream());
+        this.ignoreWhitespace = ignoreWhitespace;
+        parse();
+    }
+    
+    public void mark() {
+        stack.push( new State() );
+    }
+    
+    public void dropMark()
+    {
+        if ( stack.isEmpty() ) {
+            throw new IllegalStateException("Stack empty?");
+        }
+        stack.pop();
+    }     
+    
+    public void reset() throws IOException 
+    {
+        if ( stack.isEmpty() ) {
+            throw new IllegalStateException("Stack empty?");
+        }
+        stack.pop().apply();
+    }    
+    
+    public boolean eof() {
+        return tokens.isEmpty() && atEOF;
+    }
+    
+    public Token read(TokenType t) throws IOException
+    {
+        if ( ! peek(t) ) {
+            throw new ParseException("Expected "+t+" but got "+peek(),peek());
+        }
+        return read();
+    }
+    
+    public int currentOffset() throws IOException 
+    {
+        if ( tokens.isEmpty() ) {
+            parse();
+        }
+        return tokens.get(0).getOffset();
+    }
+    
+    public Token read() throws IOException
+    {
+        if ( tokens.isEmpty() ) {
+            parse();
+        }
+        Token result= tokens.remove(0);
+        if ( ! atEOF ) {
+            parse();
+        }
+        previousToken = result;
+        return result;
+    }
+    
+    
+    public boolean peek(TokenType t) throws IOException
+    {
+        return peek().hasType(t);
+    }
+    
+    public Token peek() throws IOException
+    {
+        if ( tokens.isEmpty() ) {
+            parse();
+        }
+        return tokens.get(0);
+    }    
+    
+    private boolean checkEOF() throws EOFException 
+    {
+    	boolean result = false;
+        if ( scanner.eof() ) 
+        {
+        	result = true;
+            if ( atEOF ) {
+                throw new EOFException();
+            }
+            atEOF = true;
+        }
+        return result;
+    }
+    
+    private void parse() throws IOException
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 97 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.